include/qore/ExceptionSink.h

```diff
@@ -309,30 +309,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
 }
 
-// returns a custom Qore program location for external modules to generate runtime exceptions with the source location
+//! returns a custom Qore program location for external modules to generate runtime exceptions with the source location
 class QoreExternalProgramLocationWrapper {
 public:
+    //! empty constructor; use set() to set the location
     DLLEXPORT QoreExternalProgramLocationWrapper();
 
+    //! copy ctor
+    DLLEXPORT QoreExternalProgramLocationWrapper(const QoreExternalProgramLocationWrapper&);
+
+    //! move ctor
+    DLLEXPORT QoreExternalProgramLocationWrapper(QoreExternalProgramLocationWrapper&&);
+
+    //! constructor setting the source location
     DLLEXPORT QoreExternalProgramLocationWrapper(const char* file, int start_line, int end_line,
-        const char* source = nullptr, int offset = 0);
+        const char* source = nullptr, int offset = 0, const char* lang = nullptr);
 
+    //! destructor; frees memory
     DLLEXPORT ~QoreExternalProgramLocationWrapper();
 
+    //! sets the program source location
     DLLEXPORT void set(const char* file, int start_line, int end_line,
-        const char* source = nullptr, int offset = 0);
+        const char* source = nullptr, int offset = 0, const char* lang = nullptr);
 
+    //! returns the source location
     DLLEXPORT const QoreProgramLocation& get() const {
         return *loc;
     }
 
+    //! returns the file name
+    DLLEXPORT const char* getFile() const {
+        return file_str.c_str();
+    }
+
+    //! returns the source
+    DLLEXPORT const char* getSource() const {
+        return source_str.c_str();
+    }
+
+    //! returns the language
+    DLLEXPORT const char* getLanguage() const {
+        return lang_str.c_str();
+    }
+
+    //! returns the start line
+    DLLEXPORT int getStartLine() const;
+
+    //! returns the start line
+    DLLEXPORT int getEndLine() const;
+
 private:
     // save strings for exceptions in case they are epheremal when this object is created
     std::string file_str;
     std::string source_str;
+    std::string lang_str;
 
     // actual exception location
     QoreProgramLocation* loc;
 };
 
+//! Stack location element abstract class
+/** @since %Qore 0.9
+*/
+class QoreStackLocation {
+public:
+    //! constructor
+    DLLLOCAL QoreStackLocation();
+
+    //! copy ctor
+    DLLLOCAL QoreStackLocation(const QoreStackLocation&) = default;
+
+    //! move ctor
+    DLLLOCAL QoreStackLocation(QoreStackLocation&&) = default;
+
+    //! virtual destructor
+    DLLLOCAL virtual ~QoreStackLocation() = default;
+
+    //! no assignment operator
+    DLLLOCAL QoreStackLocation& operator=(const QoreStackLocation&) = delete;
+
+    //! no move assignment operator
+    DLLLOCAL QoreStackLocation& operator=(QoreStackLocation&&) = delete;
+
+    //! called when pushed on the stack to set the next location
+    DLLLOCAL void setNext(const QoreStackLocation* next) {
```

Comment:
I don't understand the problem or how to solve it - can you please give me more info?
`next` is a pointer that must be valid while the current object is valid, because it's the current last stack location.  It will only disappear after the current item has been deleted.  If there is a better way to do this, please let me know.

Reply:
who is the owner of the pointer and when it gets deleted? because the API of this method takes a const* - readonly pointer somewhere - there are no guaranties that the pointer will be valid and still the method saves it into a member and accesses it later. So either the `QoreStackLocation` class is going to be owner of it, then I would do `setNext(std::unique_ptr<const QoreStackLocation> next)` or is not, and then it must be internally copied.

otherwise someone can do
```
QoreStackLocation loc1;
{
  QoreStackLocation loc2;
  loc1.setNext(&loc2);
}
loc1.next().smth - FAIL
```

Reply:
the same for the member right, it's just a pointer without knowing who owns it, I would go with unique or shared_ptr probably...?
Or as I said in a different comment, maybe the StackLocation should hold only info about the location itself and the linked list should be done by std::list - single responsibility for the object.